File: android/app/src/main/java/com/cashu/me/Core/Protocols/StorageProtocol.kt
package com.cashu.me.Core.Protocols

interface KeyValueStore {
    fun string(key: String): String?
    fun putString(key: String, value: String?)
    fun remove(key: String)
    fun removePrefix(prefix: String)
}

interface SecureStorage {
    fun loadString(key: String): String?
    fun saveString(key: String, value: String)
    fun delete(key: String)
    fun deletePrefix(prefix: String) = Unit
    fun contains(key: String): Boolean
}

fun SecureStorage.saveMnemonic(mnemonic: String) {
    saveString(StorageKeys.secureWalletMnemonic, mnemonic)
}

fun SecureStorage.loadMnemonic(): String? = loadString(StorageKeys.secureWalletMnemonic)

fun SecureStorage.deleteMnemonic() {
    delete(StorageKeys.secureWalletMnemonic)
}

fun SecureStorage.hasMnemonic(): Boolean = contains(StorageKeys.secureWalletMnemonic)

fun SecureStorage.saveNostrPrivateKey(privateKeyHex: String) {
    saveString(StorageKeys.secureNostrPrivateKey, privateKeyHex)
}

fun SecureStorage.loadNostrPrivateKey(): String? = loadString(StorageKeys.secureNostrPrivateKey)

fun SecureStorage.deleteNostrPrivateKey() {
    delete(StorageKeys.secureNostrPrivateKey)
}

fun SecureStorage.hasNostrPrivateKey(): Boolean = contains(StorageKeys.secureNostrPrivateKey)

object StorageKeys {
    const val walletDataPrefix = "wallet."
    const val walletRestoreIncomplete = "wallet.restoreIncomplete"
    const val settingsDataPrefix = "settings."
    const val npcDataPrefix = "npc."
    const val priceDataPrefix = "price."

    const val walletMints = "wallet.mints"
    const val walletRemovedMintUrls = "wallet.removedMintUrls"
    const val walletActiveMintUrl = "wallet.activeMintUrl"
    const val walletBalancesByUnit = "wallet.balancesByUnit"
    const val walletPendingReceiveTokens = "wallet.pendingReceiveTokens"
    const val walletTransactions = "wallet.transactions"
    const val walletSavedTokens = "wallet.savedTokens"
    const val walletPaymentPreimages = "wallet.paymentPreimages"
    const val walletMintQuoteTimestamps = "wallet.mintQuoteTimestamps"
    const val walletMintQuoteSchedules = "wallet.mintQuoteSchedules.v1"
    const val walletProcessedNPCQuotes = "wallet.processedNPCQuotes"
    const val walletProcessedCashuRequests = "wallet.processedCashuRequests"
    const val walletProcessedNip17GiftWraps = "wallet.processedNip17GiftWraps"
    const val cashuRequests = "cashuRequests.v1"
    const val cashuRequestsCurrentId = "cashuRequests.currentId.v1"
    const val cashuRequestsProcessedNip17Ids = "cashuRequests.processedNIP17Ids.v1"

    // Retired by the CDK 0.18 upgrade: pending/claimed send tracking, async-melt
    // tracking, and melt fee records — CDK now owns pending-transaction
    // lifecycle state. They stay on the boundary wipe list so wallet deletion
    // still covers installs that never ran the purge.
    const val retiredWalletPendingTokens = "wallet.pendingTokens"
    const val retiredWalletClaimedTokens = "wallet.claimedTokens"
    const val retiredWalletMeltQuoteFees = "wallet.meltQuoteFees"
    const val retiredWalletPendingMeltQuotes = "wallet.pendingMeltQuotes"

    const val settingsUseBitcoinSymbol = "settings.useBitcoinSymbol"
    const val settingsShowFiatBalance = "settings.showFiatBalance"
    const val settingsBitcoinPriceCurrency = "settings.bitcoinPriceCurrency"
    const val settingsCheckPendingOnStartup = "settings.checkPendingOnStartup"
    const val settingsCheckSentTokens = "settings.checkSentTokens"
    const val settingsAutoPasteEcashReceive = "settings.autoPasteEcashReceive"
    const val settingsUseWebsockets = "settings.useWebsockets"
    const val settingsEnablePaymentRequests = "settings.enablePaymentRequests"
    const val settingsReceivePaymentRequestsAutomatically = "settings.receivePaymentRequestsAutomatically"
    const val settingsShowP2PKButtonInDrawer = "settings.showP2PKButtonInDrawer"
    const val settingsP2PKKeys = "settings.p2pkKeys"
    const val settingsP2PKPendingDeletionIds = "settings.p2pkPendingDeletionIds"
    const val settingsCheckIncomingInvoices = "settings.checkIncomingInvoices"
    const val settingsPeriodicallyCheckIncomingInvoices = "settings.periodicallyCheckIncomingInvoices"
    const val settingsNostrRelays = "settings.nostrRelays"
    const val settingsNostrSignerType = "settings.nostrSignerType"
    const val settingsNostrMintBackupEnabled = "settings.nostrMintBackupEnabled"
    const val walletNostrMintBackupLastBackupDate = "wallet.nostrMintBackup.lastBackupDate"
    const val settingsAmountDisplayPrimary = "settings.amountDisplayPrimary"
    const val settingsHomeBalancePrimary = "settings.homeBalancePrimary"
    const val settingsHomeBalanceUnit = "settings.homeBalanceUnit"
    const val settingsSentryEnabled = "settings.sentryEnabled"
    const val settingsAppLockEnabled = "settings.appLockEnabled"

    // Onboarding completion marker. Written `false` when a wallet is installed
    // during first-launch onboarding and `true` only once onboarding is fully
    // passed (past the first-mint screen via Continue or Skip, or a finished
    // restore); absent on installs that predate the marker — those are
    // grandfathered to completed at launch.
    const val onboardingCompleted = "cashu.local.onboardingCompleted"

    const val npcEnabled = "npc.enabled"
    const val npcAutomaticClaim = "npc.automaticClaim"
    const val npcSelectedMint = "npc.selectedMint"
    const val npcLastCheck = "npc.lastCheck"

    const val nwcDataPrefix = "nwc."
    const val nwcEnabled = "nwc.enabled"
    const val nwcSelectedMint = "nwc.selectedMint"
    const val nwcBudgetSats = "nwc.budgetSats"

    const val priceEnabled = "price.enabled"
    const val priceCurrencyCode = "price.currencyCode"
    const val priceCachedBTC = "price.cachedBTC"
    const val priceCachedBTCDate = "price.cachedBTCDate"

    fun priceCachedBTC(currency: String) = "$priceCachedBTC.${currency.uppercase()}"
    fun priceCachedBTCDate(currency: String) = "$priceCachedBTCDate.${currency.uppercase()}"

    const val secureWalletMnemonic = "wallet_mnemonic"
    const val secureNostrPrivateKey = "nostr_private_key"
    const val secureNwcConnectionUri = "nwc_connection_uri"

    // Android's pre-CDK settings-only NWC prototype. Its random keys cannot be
    // restored by CDK's deterministic service, so they are removed on upgrade.
    const val legacySettingsEnableNwc = "settings.enableNWC"
    const val legacySettingsNwcConnections = "settings.nwcConnections"
    const val legacySecureNwcPrefix = "settings.nwc."

    val retiredWalletKeys = setOf(
        retiredWalletPendingTokens,
        retiredWalletClaimedTokens,
        retiredWalletMeltQuoteFees,
        retiredWalletPendingMeltQuotes,
    )

    val walletBoundaryKeys = setOf(
        walletMints,
        walletRemovedMintUrls,
        walletActiveMintUrl,
        walletBalancesByUnit,
        walletPendingReceiveTokens,
        walletTransactions,
        walletSavedTokens,
        walletPaymentPreimages,
        walletMintQuoteTimestamps,
        walletMintQuoteSchedules,
        walletProcessedNPCQuotes,
        walletProcessedCashuRequests,
        walletProcessedNip17GiftWraps,
        cashuRequests,
        cashuRequestsCurrentId,
        settingsP2PKKeys,
        settingsP2PKPendingDeletionIds,
        npcEnabled,
        npcAutomaticClaim,
        npcSelectedMint,
        npcLastCheck,
        nwcEnabled,
        nwcSelectedMint,
        nwcBudgetSats,
    ) + retiredWalletKeys
}

File: ios/CashuWallet/Core/Protocols/StorageProtocol.swift
import Foundation

// MARK: - Storage Protocol

/// Protocol for persistent storage operations.
/// Abstracts storage implementation to allow for different backends (UserDefaults, SQLite, Keychain, etc.)
protocol StorageProtocol {
    /// Store a value for a key
    func set<T: Codable>(_ value: T, forKey key: String) throws
    
    /// Retrieve a value for a key
    func get<T: Codable>(forKey key: String) throws -> T?
    
    /// Remove a value for a key
    func remove(forKey key: String) throws
    
    /// Check if a key exists
    func exists(forKey key: String) -> Bool
    
    /// Get all keys with a given prefix
    func keys(withPrefix prefix: String) -> [String]
}

// MARK: - Secure Storage Protocol

/// Protocol for secure storage (Keychain)
protocol SecureStorageProtocol {
    /// Store a secret securely
    func saveSecret(_ secret: String, forKey key: String) throws
    
    /// Retrieve a secret
    func loadSecret(forKey key: String) throws -> String?
    
    /// Delete a secret
    func deleteSecret(forKey key: String) throws
    
    /// Check if a secret exists
    func hasSecret(forKey key: String) -> Bool
}

// MARK: - UserDefaults Storage Implementation

/// Storage implementation using UserDefaults
final class UserDefaultsStorage: StorageProtocol {
    private let defaults: UserDefaults
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()
    
    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }
    
    func set<T: Codable>(_ value: T, forKey key: String) throws {
        let data = try encoder.encode(value)
        defaults.set(data, forKey: key)
    }
    
    func get<T: Codable>(forKey key: String) throws -> T? {
        if let data = defaults.data(forKey: key) {
            return try decoder.decode(T.self, from: data)
        }

        // Legacy compatibility for values previously written directly to UserDefaults.
        return defaults.object(forKey: key) as? T
    }
    
    func remove(forKey key: String) throws {
        defaults.removeObject(forKey: key)
    }
    
    func exists(forKey key: String) -> Bool {
        defaults.object(forKey: key) != nil
    }
    
    func keys(withPrefix prefix: String) -> [String] {
        defaults.dictionaryRepresentation().keys.filter { $0.hasPrefix(prefix) }
    }
}

// MARK: - Storage Keys

/// Centralized storage key definitions
enum StorageKeys {
    static let walletDataPrefix = "wallet."
    static let npcDataPrefix = "npc."
    static let nwcDataPrefix = "nwc."

    // Wallet
    static let mints = "wallet.mints"
    static let removedMintUrls = "wallet.removedMintUrls"
    static let activeMintUrl = "wallet.activeMintUrl"
    static let balancesByUnit = "wallet.balancesByUnit"
    static let pendingReceiveTokens = "wallet.pendingReceiveTokens"
    static let savedTokens = "wallet.savedTokens"
    static let paymentPreimages = "wallet.paymentPreimages"
    static let mintQuoteTimestamps = "wallet.mintQuoteTimestamps"
    static let mintQuoteSchedules = "wallet.mintQuoteSchedules.v1"
    static let mintKeysetRefreshTimestamps = "wallet.mintKeysetRefreshTimestamps"
    static let processedNPCQuotes = "wallet.processedNPCQuotes"
    static let nostrMintBackupLastBackupDate = "wallet.nostrMintBackup.lastBackupDate"

    // Onboarding completion marker. Written `false` when a wallet is installed
    // and `true` only when onboarding is fully passed; absent on installs that
    // predate the marker (treated as completed at launch).
    static let onboardingCompleted = "cashu.local.onboardingCompleted"

    // Cashu Requests (receive intents shown in History). Key names predate the
    // `wallet.` prefix convention and must stay stable for existing installs,
    // so they are wallet-scoped via `walletDataKeys` membership instead.
    static let cashuRequests = "cashuRequests.v1"
    static let cashuRequestsCurrentId = "cashuRequests.currentId.v1"
    static let cashuRequestsProcessedNIP17Ids = "cashuRequests.nip17.processedIds.v1"

    // Settings
    static let useBitcoinSymbol = "settings.useBitcoinSymbol"
    static let showFiatBalance = "settings.showFiatBalance"
    static let bitcoinPriceCurrency = "settings.bitcoinPriceCurrency"
    static let checkSentTokens = "settings.checkSentTokens"
    static let autoPasteEcashReceive = "settings.autoPasteEcashReceive"
    static let useWebsockets = "settings.useWebsockets"
    static let enablePaymentRequests = "settings.enablePaymentRequests"
    static let receivePaymentRequestsAutomatically = "settings.receivePaymentRequestsAutomatically"
    static let showP2PKButtonInDrawer = "settings.showP2PKButtonInDrawer"
    static let p2pkKeys = "settings.p2pkKeys"
    static let p2pkPendingDeletionIDs = "settings.p2pkPendingDeletionIDs"
    static let checkIncomingInvoices = "settings.checkIncomingInvoices"
    static let periodicallyCheckIncomingInvoices = "settings.periodicallyCheckIncomingInvoices"
    static let nostrRelays = "settings.nostrRelays"
    static let nostrSignerType = "settings.nostrSignerType"
    static let nostrMintBackupEnabled = "settings.nostrMintBackupEnabled"
    static let amountDisplayPrimary = "settings.amountDisplayPrimary"
    static let homeBalancePrimary = "settings.homeBalancePrimary"
    static let appLockEnabled = "settings.appLockEnabled"
    static let sentryEnabled = "settings.sentryEnabled"

    /// Keys retired by the CDK 0.18 upgrade: pending/claimed send tracking,
    /// async-melt tracking, melt fee records, and the long-vestigial
    /// transaction cache. CDK now owns pending-transaction lifecycle state, so
    /// these stores are dead weight (and hold token strings at rest). They stay
    /// on the wipe lists so wallet deletion still covers installs that never
    /// ran the purge.
    enum Retired {
        static let pendingTokens = "wallet.pendingTokens"
        static let claimedTokens = "wallet.claimedTokens"
        static let transactions = "wallet.transactions"
        static let meltQuoteFees = "wallet.meltQuoteFees"
        static let pendingMeltQuotes = "wallet.pendingMeltQuotes"

        static let all = [
            pendingTokens,
            claimedTokens,
            transactions,
            meltQuoteFees,
            pendingMeltQuotes,
            Legacy.pendingTokens,
            Legacy.claimedTokens
        ]
    }

    enum Legacy {
        static let mints = "savedMints"
        static let pendingTokens = "pendingTokens"
        static let pendingReceiveTokens = "pendingReceiveTokens"
        static let claimedTokens = "claimedTokens"
        static let savedTokens = "savedTokens"
        static let paymentPreimages = "paymentPreimages"
        static let mintQuoteTimestamps = "mintQuoteTimestamps"
        static let useBitcoinSymbol = "useBitcoinSymbol"
        static let showFiatBalance = "showFiatBalance"
        static let bitcoinPriceCurrency = "bitcoinPriceCurrency"
        static let checkSentTokens = "checkSentTokens"
        static let autoPasteEcashReceive = "autoPasteEcashReceive"
        static let useWebsockets = "useWebsockets"
        static let enablePaymentRequests = "enablePaymentRequests"
        static let receivePaymentRequestsAutomatically = "receivePaymentRequestsAutomatically"
        static let showP2PKButtonInDrawer = "showP2PKButtonInDrawer"
        static let p2pkKeys = "p2pkKeys"
        static let checkIncomingInvoices = "checkIncomingInvoices"
        static let periodicallyCheckIncomingInvoices = "periodicallyCheckIncomingInvoices"
        static let nostrRelays = "nostrRelays"
        static let nostrSignerType = "nostr_signer_type"
        static let priceEnabled = "priceServiceEnabled"
        static let priceCurrencyCode = "priceServiceCurrencyCode"
        static let cachedBTCPrice = "cachedBTCPrice"
        static let cachedBTCPriceDate = "cachedBTCPriceDate"

        static func cachedBTCPrice(currency: String) -> String {
            "cachedBTCPrice.\(currency.uppercased())"
        }

        static func cachedBTCPriceDate(currency: String) -> String {
            "cachedBTCPriceDate.\(currency.uppercased())"
        }
    }
    
    // NPC
    static let npcEnabled = "npc.enabled"
    static let npcAutomaticClaim = "npc.automaticClaim"
    static let npcSelectedMint = "npc.selectedMint"
    static let npcLastCheck = "npc.lastCheck"

    // NWC (Nostr Wallet Connect)
    static let nwcEnabled = "nwc.enabled"
    static let nwcConnectionUri = "nwc.connectionUri"
    static let nwcSelectedMint = "nwc.selectedMint"
    static let nwcBudgetSats = "nwc.budgetSats"
    
    // Price
    static let priceEnabled = "price.enabled"
    static let priceCurrencyCode = "price.currencyCode"
    static let cachedBTCPrice = "price.cachedBTC"
    static let cachedBTCPriceDate = "price.cachedBTCDate"

    static func cachedBTCPrice(currency: String) -> String {
        "price.cachedBTC.\(currency.uppercased())"
    }

    static func cachedBTCPriceDate(currency: String) -> String {
        "price.cachedBTCDate.\(currency.uppercased())"
    }

    static let walletDataKeys = [
        mints,
        removedMintUrls,
        activeMintUrl,
        balancesByUnit,
        pendingReceiveTokens,
        savedTokens,
        paymentPreimages,
        mintQuoteTimestamps,
        mintQuoteSchedules,
        mintKeysetRefreshTimestamps,
        processedNPCQuotes,
        nostrMintBackupLastBackupDate,
        cashuRequests,
        cashuRequestsCurrentId,
        cashuRequestsProcessedNIP17Ids,
        Retired.pendingTokens,
        Retired.claimedTokens,
        Retired.transactions,
        Retired.meltQuoteFees,
        Retired.pendingMeltQuotes
    ]

    static let walletDataLegacyKeys = [
        Legacy.mints,
        Legacy.pendingTokens,
        Legacy.pendingReceiveTokens,
        Legacy.claimedTokens,
        Legacy.savedTokens,
        Legacy.paymentPreimages,
        Legacy.mintQuoteTimestamps
    ]

    static let walletScopedSettingsKeys = [
        enablePaymentRequests,
        receivePaymentRequestsAutomatically,
        showP2PKButtonInDrawer,
        p2pkKeys,
        p2pkPendingDeletionIDs,
        nostrSignerType,
        nostrMintBackupEnabled,
        npcEnabled,
        npcAutomaticClaim,
        npcSelectedMint,
        npcLastCheck,
        nwcEnabled,
        nwcConnectionUri,
        nwcSelectedMint,
        nwcBudgetSats
    ]

    static let walletScopedSettingsLegacyKeys = [
        Legacy.enablePaymentRequests,
        Legacy.receivePaymentRequestsAutomatically,
        Legacy.showP2PKButtonInDrawer,
        Legacy.p2pkKeys,
        Legacy.nostrSignerType
    ]

    static var walletBoundaryKeys: [String] {
        walletDataKeys + walletDataLegacyKeys + walletScopedSettingsKeys + walletScopedSettingsLegacyKeys
            + [onboardingCompleted, ICloudRestoreState.incompleteKey]
    }
    
    // Keychain (Secure Storage)
    enum Secure {
        static let mnemonic = "wallet_mnemonic"
        static let nostrPrivateKey = "nostr_private_key"
    }
}
